File: src/core/acp/provider-adapter/trace-recorder.ts
/**
 * Trace Recorder
 *
 * Records traces from normalized session updates.
 * Handles deferred input patterns using a pending tool calls buffer.
 */

import {
  createTraceRecord,
  withConversation,
  withTool,
  withVcs,
  recordTrace,
  extractFilesFromToolCall,
  getVcsContextLight,
} from "@/core/trace";
import type { NormalizedSessionUpdate, NormalizedToolCall } from "./types";

/**
 * Pending tool call waiting for input.
 */
interface PendingToolCall {
  toolCallId: string;
  name: string;
  title?: string;
  sessionId: string;
  cwd: string;
  provider: string;
  traced: boolean;
}

/**
 * TraceRecorder handles recording traces from normalized session updates.
 * It manages pending tool calls for providers that send input in updates.
 */
export class TraceRecorder {
  /** Buffer for pending tool calls awaiting input */
  private pendingToolCalls = new Map<string, PendingToolCall>();
  /** Buffer for accumulating message chunks */
  private messageBuffer = new Map<string, string>();
  /** Buffer for accumulating thought chunks */
  private thoughtBuffer = new Map<string, string>();

  /**
   * Record a trace from a normalized session update.
   *
   * @param update - The normalized session update
   * @param cwd - Working directory for trace storage
   */
  recordFromUpdate(update: NormalizedSessionUpdate, cwd: string): void {
    const { sessionId, provider, eventType } = update;

    switch (eventType) {
      case "tool_call":
        this.handleToolCall(update, cwd);
        break;

      case "tool_call_update":
        this.handleToolCallUpdate(update, cwd);
        break;

      case "agent_message":
        this.handleAgentMessage(update, cwd);
        break;

      case "agent_thought":
        this.handleAgentThought(update, cwd);
        break;

      case "user_message":
        this.handleUserMessage(update, cwd);
        break;

      case "turn_complete":
        this.flushBuffers(sessionId, cwd, provider);
        break;

      // Errors and other types are not traced
    }
  }

  private handleToolCall(update: NormalizedSessionUpdate, cwd: string): void {
    const { sessionId, provider, toolCall } = update;
    if (!toolCall) return;

    if (toolCall.inputFinalized) {
      // Input is ready, record immediately
      this.recordToolCallTrace(sessionId, provider, toolCall, cwd);
    } else {
      // Input is deferred, store in pending
      this.pendingToolCalls.set(toolCall.toolCallId, {
        toolCallId: toolCall.toolCallId,
        name: toolCall.name,
        title: toolCall.title,
        sessionId,
        cwd,
        provider,
        traced: false,
      });
    }
  }

  private handleToolCallUpdate(update: NormalizedSessionUpdate, cwd: string): void {
    const { sessionId, provider, toolCall } = update;
    if (!toolCall) return;

    // Check if this update provides deferred input
    const pending = this.pendingToolCalls.get(toolCall.toolCallId);
    if (pending && !pending.traced && toolCall.inputFinalized && toolCall.input) {
      // Record the tool_call trace with the now-available input
      const finalToolCall: NormalizedToolCall = {
        ...toolCall,
        name: toolCall.name || pending.name,
        title: toolCall.title || pending.title,
      };
      this.recordToolCallTrace(sessionId, provider, finalToolCall, cwd);
      pending.traced = true;
    }

    // Record tool_result if complete
    const isComplete = toolCall.status === "completed" || toolCall.status === "failed";
    if (isComplete) {
      this.recordToolResultTrace(sessionId, provider, toolCall, cwd);
      // Clean up pending
      this.pendingToolCalls.delete(toolCall.toolCallId);
    }
  }

  private handleAgentMessage(update: NormalizedSessionUpdate, cwd: string): void {
    const { sessionId, provider, message } = update;
    if (!message) return;

    if (message.isChunk) {
      // Accumulate chunks
      const existing = this.messageBuffer.get(sessionId) ?? "";
      const accumulated = existing + message.content;
      this.messageBuffer.set(sessionId, accumulated);

      // Trace when buffer reaches threshold
      if (accumulated.length >= 100) {
        this.recordAgentMessageTrace(sessionId, provider, accumulated, cwd);
        this.messageBuffer.set(sessionId, "");
      }
    } else {
      // Complete message, trace immediately
      this.recordAgentMessageTrace(sessionId, provider, message.content, cwd);
    }
  }

  private handleAgentThought(update: NormalizedSessionUpdate, cwd: string): void {
    const { sessionId, provider, message } = update;
    if (!message) return;

    if (message.isChunk) {
      const existing = this.thoughtBuffer.get(sessionId) ?? "";
      const accumulated = existing + message.content;
      this.thoughtBuffer.set(sessionId, accumulated);

      if (accumulated.length >= 100) {
        this.recordAgentThoughtTrace(sessionId, provider, accumulated, cwd);
        this.thoughtBuffer.set(sessionId, "");
      }
    } else {
      this.recordAgentThoughtTrace(sessionId, provider, message.content, cwd);
    }
  }

  private handleUserMessage(update: NormalizedSessionUpdate, cwd: string): void {
    const { sessionId, provider, message } = update;
    if (!message) return;

    let trace = createTraceRecord(sessionId, "user_message", { provider });
    trace = withConversation(trace, {
      role: "user",
      contentPreview: message.content.slice(0, 200),
      fullContent: message.content,
    });
    recordTrace(cwd, trace);
  }

  private flushBuffers(sessionId: string, cwd: string, provider: string): void {
    // Flush message buffer
    const message = this.messageBuffer.get(sessionId);
    if (message && message.length > 0) {
      this.recordAgentMessageTrace(sessionId, provider, message, cwd);
      this.messageBuffer.set(sessionId, "");
    }

    // Flush thought buffer
    const thought = this.thoughtBuffer.get(sessionId);
    if (thought && thought.length > 0) {
      this.recordAgentThoughtTrace(sessionId, provider, thought, cwd);
      this.thoughtBuffer.set(sessionId, "");
    }
  }

  private recordToolCallTrace(
    sessionId: string,
    provider: string,
    toolCall: NormalizedToolCall,
    cwd: string
  ): void {
    let trace = createTraceRecord(sessionId, "tool_call", { provider });
    trace = withTool(trace, {
      name: toolCall.name,
      toolCallId: toolCall.toolCallId,
      status: "running",
      input: toolCall.input,
    });

    // Extract file ranges from tool parameters
    const files = extractFilesFromToolCall(toolCall.name, toolCall.input);
    if (files.length > 0) {
      trace = { ...trace, files };
    }

    // Add VCS context
    const vcs = getVcsContextLight(cwd);
    if (vcs) {
      trace = withVcs(trace, vcs);
    }

    recordTrace(cwd, trace);
  }

  private recordToolResultTrace(
    sessionId: string,
    provider: string,
    toolCall: NormalizedToolCall,
    cwd: string
  ): void {
    let trace = createTraceRecord(sessionId, "tool_result", { provider });
    trace = withTool(trace, {
      name: toolCall.name,
      toolCallId: toolCall.toolCallId,
      status: toolCall.status,
      output: toolCall.output as string | undefined,
    });
    recordTrace(cwd, trace);
  }

  private recordAgentMessageTrace(
    sessionId: string,
    provider: string,
    content: string,
    cwd: string
  ): void {
    let trace = createTraceRecord(sessionId, "agent_message", { provider });
    trace = withConversation(trace, {
      role: "assistant",
      contentPreview: content.slice(0, 200),
      fullContent: content,
    });
    recordTrace(cwd, trace);
  }

  private recordAgentThoughtTrace(
    sessionId: string,
    provider: string,
    content: string,
    cwd: string
  ): void {
    let trace = createTraceRecord(sessionId, "agent_thought", { provider });
    trace = withConversation(trace, {
      role: "assistant",
      contentPreview: content.slice(0, 200),
      fullContent: content,
    });
    recordTrace(cwd, trace);
  }

  /**
   * Flush and record any buffered content for a session.
   * Call this when a prompt completes.
   *
   * @param sessionId - Session to flush
   * @param cwd - Working directory for trace storage
   * @param provider - Provider name for trace metadata
   */
  flushSession(sessionId: string, cwd: string, provider: string): void {
    this.flushBuffers(sessionId, cwd, provider);
  }

  /**
   * Clean up session data when session ends.
   */
  cleanupSession(sessionId: string): void {
    this.messageBuffer.delete(sessionId);
    this.thoughtBuffer.delete(sessionId);
    // Clean up any pending tool calls for this session
    for (const [toolCallId, pending] of this.pendingToolCalls) {
      if (pending.sessionId === sessionId) {
        this.pendingToolCalls.delete(toolCallId);
      }
    }
  }
}

File: crates/routa-core/src/acp/provider_adapter/trace_recorder.rs
//! Trace Recorder
//!
//! Records traces from normalized session updates.
//! Handles deferred input patterns using a pending tool calls buffer.

use std::collections::HashMap;

use crate::trace::{
    TraceRecord, TraceWriter, TraceEventType, Contributor, TraceConversation, TraceTool,
};
use super::types::{
    NormalizedSessionUpdate, NormalizedEventType, NormalizedToolCall, ToolStatus,
};

/// Pending tool call waiting for input.
#[derive(Debug)]
struct PendingToolCall {
    tool_call_id: String,
    name: String,
    title: Option<String>,
    #[allow(dead_code)]
    provider: String,
    traced: bool,
}

/// TraceRecorder handles recording traces from session updates.
/// It manages pending tool calls for providers that send input in updates.
pub struct TraceRecorder {
    /// Buffer for pending tool calls awaiting input
    pending_tool_calls: HashMap<String, PendingToolCall>,
    /// Buffer for accumulating message chunks
    message_buffer: HashMap<String, String>,
    /// Buffer for accumulating thought chunks
    thought_buffer: HashMap<String, String>,
}

impl TraceRecorder {
    pub fn new() -> Self {
        Self {
            pending_tool_calls: HashMap::new(),
            message_buffer: HashMap::new(),
            thought_buffer: HashMap::new(),
        }
    }

    /// Record a trace from a normalized session update.
    pub async fn record_from_update(
        &mut self,
        update: &NormalizedSessionUpdate,
        cwd: &str,
    ) {
        match update.event_type {
            NormalizedEventType::ToolCall => self.handle_tool_call(update, cwd).await,
            NormalizedEventType::ToolCallUpdate => self.handle_tool_call_update(update, cwd).await,
            NormalizedEventType::AgentMessage => self.handle_agent_message(update, cwd).await,
            NormalizedEventType::AgentThought => self.handle_agent_thought(update, cwd).await,
            NormalizedEventType::UserMessage => self.handle_user_message(update, cwd).await,
            NormalizedEventType::TurnComplete => self.flush_buffers(&update.session_id, cwd, &update.provider).await,
            NormalizedEventType::PlanUpdate | NormalizedEventType::Error => {}
        }
    }

    async fn handle_tool_call(&mut self, update: &NormalizedSessionUpdate, cwd: &str) {
        let Some(tool_call) = &update.tool_call else { return };

        if tool_call.input_finalized {
            // Input is ready, record immediately
            self.record_tool_call_trace(&update.session_id, &update.provider, tool_call, cwd).await;
        } else {
            // Input is deferred, store in pending
            self.pending_tool_calls.insert(tool_call.tool_call_id.clone(), PendingToolCall {
                tool_call_id: tool_call.tool_call_id.clone(),
                name: tool_call.name.clone(),
                title: tool_call.title.clone(),
                provider: update.provider.clone(),
                traced: false,
            });
        }
    }

    async fn handle_tool_call_update(&mut self, update: &NormalizedSessionUpdate, cwd: &str) {
        let Some(tool_call) = &update.tool_call else { return };

        // Check if this update provides deferred input - extract data first to avoid borrow issues
        let should_trace_call = {
            if let Some(pending) = self.pending_tool_calls.get(&tool_call.tool_call_id) {
                !pending.traced && tool_call.input_finalized && tool_call.input.is_some()
            } else {
                false
            }
        };

        let final_call_data = if should_trace_call {
            let pending = self.pending_tool_calls.get(&tool_call.tool_call_id).unwrap();
            Some(NormalizedToolCall {
                tool_call_id: tool_call.tool_call_id.clone(),
                name: if tool_call.name.is_empty() { pending.name.clone() } else { tool_call.name.clone() },
                title: tool_call.title.clone().or_else(|| pending.title.clone()),
                status: ToolStatus::Running,
                input: tool_call.input.clone(),
                output: None,
                input_finalized: true,
            })
        } else {
            None
        };

        // Now we can safely call async methods and mutate
        if let Some(final_call) = final_call_data {
            self.record_tool_call_trace(&update.session_id, &update.provider, &final_call, cwd).await;
            if let Some(pending) = self.pending_tool_calls.get_mut(&tool_call.tool_call_id) {
                pending.traced = true;
            }
        }

        // Record tool_result if complete
        if matches!(tool_call.status, ToolStatus::Completed | ToolStatus::Failed) {
            self.record_tool_result_trace(&update.session_id, &update.provider, tool_call, cwd).await;
            self.pending_tool_calls.remove(&tool_call.tool_call_id);
        }
    }

    async fn handle_agent_message(&mut self, update: &NormalizedSessionUpdate, cwd: &str) {
        let Some(message) = &update.message else { return };

        if message.is_chunk {
            let existing = self.message_buffer.entry(update.session_id.clone()).or_default();
            existing.push_str(&message.content);

            if existing.len() >= 100 {
                let content = std::mem::take(existing);
                self.record_agent_message_trace(&update.session_id, &update.provider, &content, cwd).await;
            }
        } else {
            self.record_agent_message_trace(&update.session_id, &update.provider, &message.content, cwd).await;
        }
    }

    async fn handle_agent_thought(&mut self, update: &NormalizedSessionUpdate, cwd: &str) {
        let Some(message) = &update.message else { return };

        if message.is_chunk {
            let existing = self.thought_buffer.entry(update.session_id.clone()).or_default();
            existing.push_str(&message.content);

            if existing.len() >= 100 {
                let content = std::mem::take(existing);
                self.record_agent_thought_trace(&update.session_id, &update.provider, &content, cwd).await;
            }
        } else {
            self.record_agent_thought_trace(&update.session_id, &update.provider, &message.content, cwd).await;
        }
    }

    async fn handle_user_message(&mut self, update: &NormalizedSessionUpdate, cwd: &str) {
        let Some(message) = &update.message else { return };
        
        let record = TraceRecord::new(
            &update.session_id,
            TraceEventType::UserMessage,
            Contributor::new(&update.provider, None),
        ).with_conversation(TraceConversation {
            turn: None,
            role: Some("user".to_string()),
            content_preview: Some(message.content.chars().take(200).collect()),
            full_content: Some(message.content.clone()),
        });
        let writer = TraceWriter::new(cwd);
        let _ = writer.append_safe(&record).await;
    }

    async fn flush_buffers(&mut self, session_id: &str, cwd: &str, provider: &str) {
        // Flush message buffer
        if let Some(content) = self.message_buffer.remove(session_id) {
            if !content.is_empty() {
                self.record_agent_message_trace(session_id, provider, &content, cwd).await;
            }
        }

        // Flush thought buffer
        if let Some(content) = self.thought_buffer.remove(session_id) {
            if !content.is_empty() {
                self.record_agent_thought_trace(session_id, provider, &content, cwd).await;
            }
        }
    }

    async fn record_tool_call_trace(
        &self,
        session_id: &str,
        provider: &str,
        tool_call: &NormalizedToolCall,
        cwd: &str,
    ) {
        let record = TraceRecord::new(
            session_id,
            TraceEventType::ToolCall,
            Contributor::new(provider, None),
        ).with_tool(TraceTool {
            name: tool_call.name.clone(),
            tool_call_id: Some(tool_call.tool_call_id.clone()),
            status: Some("running".to_string()),
            input: tool_call.input.clone(),
            output: None,
        });
        let writer = TraceWriter::new(cwd);
        let _ = writer.append_safe(&record).await;
    }

    async fn record_tool_result_trace(
        &self,
        session_id: &str,
        provider: &str,
        tool_call: &NormalizedToolCall,
        cwd: &str,
    ) {
        let record = TraceRecord::new(
            session_id,
            TraceEventType::ToolResult,
            Contributor::new(provider, None),
        ).with_tool(TraceTool {
            name: tool_call.name.clone(),
            tool_call_id: Some(tool_call.tool_call_id.clone()),
            status: Some(tool_call.status.as_str().to_string()),
            input: None,
            output: tool_call.output.clone(),
        });
        let writer = TraceWriter::new(cwd);
        let _ = writer.append_safe(&record).await;
    }

    async fn record_agent_message_trace(
        &self,
        session_id: &str,
        provider: &str,
        content: &str,
        cwd: &str,
    ) {
        let record = TraceRecord::new(
            session_id,
            TraceEventType::AgentMessage,
            Contributor::new(provider, None),
        ).with_conversation(TraceConversation {
            turn: None,
            role: Some("assistant".to_string()),
            content_preview: Some(content.chars().take(200).collect()),
            full_content: Some(content.to_string()),
        });
        let writer = TraceWriter::new(cwd);
        let _ = writer.append_safe(&record).await;
    }

    async fn record_agent_thought_trace(
        &self,
        session_id: &str,
        provider: &str,
        content: &str,
        cwd: &str,
    ) {
        let record = TraceRecord::new(
            session_id,
            TraceEventType::AgentThought,
            Contributor::new(provider, None),
        ).with_conversation(TraceConversation {
            turn: None,
            role: Some("assistant".to_string()),
            content_preview: Some(content.chars().take(200).collect()),
            full_content: Some(content.to_string()),
        });
        let writer = TraceWriter::new(cwd);
        let _ = writer.append_safe(&record).await;
    }

    /// Flush buffers for a session (call when prompt completes).
    pub async fn flush_session(&mut self, session_id: &str, cwd: &str, provider: &str) {
        self.flush_buffers(session_id, cwd, provider).await;
    }

    /// Clean up session data when session ends.
    pub fn cleanup_session(&mut self, session_id: &str) {
        self.message_buffer.remove(session_id);
        self.thought_buffer.remove(session_id);
        self.pending_tool_calls.retain(|_, v| v.tool_call_id != session_id);
    }
}

impl Default for TraceRecorder {
    fn default() -> Self {
        Self::new()
    }
}

